staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/unstructured_conversion_test.go

```diff
@@ -0,0 +1,465 @@
+			}()
+			outObject := testCase.convertingObject.DeepCopyObject()
+			// Convert by specifying destination object
+			err := scheme.Convert(testCase.unstructuredToConvert, outObject, nil)
+			if testCase.expectedErrFunc != nil {
+				if !testCase.expectedErrFunc(err) {
+					t.Errorf("error mismatched: %v", err)
+				}
+			}
+			assert.Equal(t, testCase.expectedConvertedObject, outObject)
+		})
+	}
+}
+
+func TestUnstructuredToGVConversion(t *testing.T) {
+	scheme, _ := test.TestScheme()
+	testCases := []struct {
+		name                    string
+		unstructuredToConvert   *unstructured.Unstructured
+		targetGV                schema.GroupVersion
+		expectPanic             bool
+		expectedErrFunc         func(err error) bool
+		expectedConvertedObject runtime.Object
+	}{
+		{
+			name:                  "convert nil unstructured to versioned will panic",
+			unstructuredToConvert: nil,
+			targetGV:              schema.GroupVersion{Group: "", Version: "v1"},
+			expectPanic:           true,
+			expectedConvertedObject: &testapigroupv1.Carp{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v1",
+					Kind:       "Carp",
+				},
+			},
+		},
+		{
+			name: "convert empty unstructured w/o gvk to versioned should fail",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			},
+			targetGV: schema.GroupVersion{Group: "", Version: "v1"},
+			expectedErrFunc: func(err error) bool {
+				return reflect.DeepEqual(err, runtime.NewMissingKindErr("unstructured object has no kind"))
+			},
+			expectedConvertedObject: nil,
+		},
+		{
+			name: "convert versioned unstructured to versioned should succeed",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+				},
+			},
+			targetGV: schema.GroupVersion{Group: "", Version: "v1"},
+			expectedConvertedObject: &testapigroupv1.Carp{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v1",
+					Kind:       "Carp",
+				},
+			},
+		},
+
+		{
+			name: "convert valid versioned unstructured to versioned should succeed",
+			unstructuredToConvert: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Carp",
+					"metadata": map[string]interface{}{
+						"creationTimestamp": nil,
+						"name":              "noxu",
+					},
+					"spec": map[string]interface{}{
+						"hostname": "example.com",
+					},
+					"status": map[string]interface{}{},
+				},
+			},
+			targetGV: schema.GroupVersion{Group: "", Version: "v1"},
+			expectedConvertedObject: &testapigroupv1.Carp{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v1",
+					Kind:       "Carp",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "noxu",
+				},
+				Spec: testapigroupv1.CarpSpec{
+					Hostname: "example.com",
+				},
+			},
+		},
+		{
+			name: "convert valid versioned unstructured to internal should succeed but ignore type meta",
```

Comment:
some interesting cases, here. dyt it's desgined?

Reply:
yes, type meta is dropped when converting to go structs, since the apiVersion/kind is inherent in the go type

Reply:
what is "internal" here? Is that the hub type (that should not be handled differently for unstructured). Trying to convert to internal should be refused, not ignored IMO.

Reply:
@sttts updated and renamed them to hub-version